Source/Core/Core/IOS/USB/Emulated/Microphone.cpp

```diff
@@ -0,0 +1,371 @@
+    return;
+  }
+
+  // TODO: Not here but rather inside the WiiSpeak device if possible?
+  StreamStart(m_sampler.DEFAULT_SAMPLING_RATE);
+}
+
+void Microphone::StreamTerminate()
+{
+  StreamStop();
+
+  if (m_cubeb_ctx)
+    m_worker.Execute([this] { m_cubeb_ctx.reset(); });
+}
+
+static void StateCallback(cubeb_stream* stream, void* user_data, cubeb_state state)
+{
+}
+
+void Microphone::StreamStart(u32 sampling_rate)
+{
+  if (!m_cubeb_ctx)
+    return;
+
+  m_worker.Execute([this, sampling_rate] {
+#ifdef ANDROID
+    JNIEnv* env = IDCache::GetEnvForThread();
+    if (jboolean result = env->CallStaticBooleanMethod(
+            IDCache::GetPermissionHandlerClass(),
+            IDCache::GetPermissionHandlerHasRecordAudioPermission(), nullptr);
+        result == JNI_FALSE)
+    {
+      env->CallStaticVoidMethod(IDCache::GetPermissionHandlerClass(),
+                                IDCache::GetPermissionHandlerRequestRecordAudioPermission(),
+                                nullptr);
+    }
+#endif
+
+    cubeb_stream_params params{};
+    params.format = CUBEB_SAMPLE_S16LE;
+    params.rate = sampling_rate;
+    params.channels = 1;
+    params.layout = CUBEB_LAYOUT_MONO;
+
+    u32 minimum_latency;
+    if (cubeb_get_min_latency(m_cubeb_ctx.get(), &params, &minimum_latency) != CUBEB_OK)
+    {
+      WARN_LOG_FMT(IOS_USB, "Error getting minimum latency");
+      minimum_latency = 16;
+    }
+
+    cubeb_devid input_device =
+        CubebUtils::GetInputDeviceById(Config::Get(Config::MAIN_WII_SPEAK_MICROPHONE));
+    if (cubeb_stream_init(m_cubeb_ctx.get(), &m_cubeb_stream, "Dolphin Emulated Wii Speak",
+                          input_device, &params, nullptr, nullptr,
+                          std::max<u32>(16, minimum_latency), DataCallback, StateCallback,
+                          this) != CUBEB_OK)
+    {
+      ERROR_LOG_FMT(IOS_USB, "Error initializing cubeb stream");
+      return;
+    }
+
+    if (cubeb_stream_start(m_cubeb_stream) != CUBEB_OK)
+    {
+      ERROR_LOG_FMT(IOS_USB, "Error starting cubeb stream");
+      return;
+    }
+
+    INFO_LOG_FMT(IOS_USB, "started cubeb stream");
+  });
+}
+
+void Microphone::StreamStop()
+{
+  if (!m_cubeb_stream)
+    return;
+
+  m_worker.Execute([this] {
+    if (cubeb_stream_stop(m_cubeb_stream) != CUBEB_OK)
+      ERROR_LOG_FMT(IOS_USB, "Error stopping cubeb stream");
+    cubeb_stream_destroy(m_cubeb_stream);
+    m_cubeb_stream = nullptr;
+  });
+}
+
+long Microphone::DataCallback(cubeb_stream* stream, void* user_data, const void* input_buffer,
+                              void* /*output_buffer*/, long nframes)
+{
+  // Skip data when core isn't running
+  if (Core::GetState(Core::System::GetInstance()) != Core::State::Running)
+    return nframes;
+
+  // Skip data when HLE Wii Speak is muted
+  // TODO: Update cubeb and use cubeb_stream_set_input_mute
+  if (Config::Get(Config::MAIN_WII_SPEAK_MUTED))
+    return nframes;
+
+  auto* mic = static_cast<Microphone*>(user_data);
```

Comment:
This function would be easier to look at if you made `DataCallback` call a non-static member function once you get the `this` pointer from `user_data`, so all the `mic->m_private_member` can just be `m_private_member`.

Reply:
Nice idea. It allowed to make the `UpdateLoudness` method private and remove the `GetSampler` method.